### PR TITLE
[#172161469] send email to DPO for each data processing request

### PR DESCRIPTION
**Pull request by AleDore**
Update user-data-processing functionality with, email integration to DPO for back-office purpose.

### CODE REVIEW

File: SendUserDataProcessingEmailActivity/handler.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+  const activityInput = errorOrActivityInput.value;
+  const { choice, email, fiscalCode } = activityInput;
+
+  const { from, title, to } = emailDefaults;
+
+  // prepare the text version of the message
+  const emailText = `Con la presente si informa che e' stata effettuata la richiesta di:
+  ${choice.toString()} dall' utente con codice fiscale ${fiscalCode}.
+  L' indirizzo e-mail dell' utente e' ${email}`;
```
Comment (gunzip) [inline SendUserDataProcessingEmailActivity/handler.ts:72]:
```suggestion
  const emailText = `Un utente di IO ha inoltrato una nuova richiesta:
tipo richiesta: ${choice.toString()}
codice fiscale: ${fiscalCode}
indirizzo email: ${email}.`;
```

File: SendUserDataProcessingEmailActivity/handler.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+  // Send email to DPO with information about user's will to down load or delete its own data
```
Comment (gunzip) [inline SendUserDataProcessingEmailActivity/handler.ts:74]:
```suggestion
  // Send an email to the DPO containing the information about the IO user's
  // choice to download or delete his own private data stored by the platform
```

File: SendUserDataProcessingEmailActivity/index.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+// Whether we're in a production environment
+const isProduction = process.env.NODE_ENV === "production";
+
+// Mailup
+const mailupUsername = getRequiredStringEnv("MAILUP_USERNAME");
+const mailupSecret = getRequiredStringEnv("MAILUP_SECRET");
+
+// Email data
+const EMAIL_TITLE = "IO - Richiesta di Download/Cancellazione Dati Utente";
```
Comment (gunzip) [inline SendUserDataProcessingEmailActivity/index.ts:17]: the email title is defined here while the text inside the activity. can we define both inside the activity function ? (so, just in case, we can put some user's data into the title as well)

File: SendUserDataProcessingEmailActivity/index.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+const mailFrom = getRequiredStringEnv("MAIL_FROM");
+const dpoMailTo = getRequiredStringEnv("DPO_MAIL_TO");
```
Comment (gunzip) [inline SendUserDataProcessingEmailActivity/index.ts:19]:
```suggestion
const dpoMailTo = getRequiredStringEnv("DPO_EMAIL_ADDRESS");
```

File: SendUserDataProcessingEmailActivity/index.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+// Whether we're in a production environment
+const isProduction = process.env.NODE_ENV === "production";
+
+// Mailup
+const mailupUsername = getRequiredStringEnv("MAILUP_USERNAME");
+const mailupSecret = getRequiredStringEnv("MAILUP_SECRET");
+
+// Email data
+const EMAIL_TITLE = "IO - Richiesta di Download/Cancellazione Dati Utente";
+const mailFrom = getRequiredStringEnv("MAIL_FROM");
```
Comment (gunzip) [inline SendUserDataProcessingEmailActivity/index.ts:18]:
```suggestion
const mailFrom = getRequiredStringEnv("DPO_EMAIL_FROM_ADDRESS");
```

Comment (gunzip) [reply SendUserDataProcessingEmailActivity/index.ts:18]: these settings all appear into the app service configuration and it would be difficult to tell what "mail_from" means by looking at the settings : )

File: SendUserDataProcessingEmailActivity/index.ts (hunk 3 above)
Comment (gunzip) [inline SendUserDataProcessingEmailActivity/index.ts:17]:
```suggestion
const emailTitle = "IO - Richiesta di Download/Cancellazione Dati Utente";
```

File: UpsertUserDataProcessing/handler.ts [hunk 6]
```diff
@@ -91,52 +94,69 @@ export function UpsertUserDataProcessingHandler(
...
 
-        return ResponseErrorQuery(
-          "Error while creating a new user data processing",
-          errorOrUpsertedUserDataProcessing.value
+    const createdOrUpdatedUserDataProcessing =
+      errorOrUpsertedUserDataProcessing.value;
+    const errorOrMaybeRetrievedProfile = await profileModel.findOneProfileByFiscalCode(
+      fiscalCode
+    );
+    if (isRight(errorOrMaybeRetrievedProfile)) {
```
Comment (gunzip) [inline UpsertUserDataProcessing/handler.ts:123]: can we log something in case of errors ?

File: UpsertUserDataProcessing/handler.ts [hunk 7]
```diff
@@ -91,52 +94,69 @@ export function UpsertUserDataProcessingHandler(
...
-
-        context.log.error(`${logPrefix}|ERROR=${body}`);
+    }
 
-        return ResponseErrorQuery(
-          "Error while creating a new user data processing",
-          errorOrUpsertedUserDataProcessing.value
+    const createdOrUpdatedUserDataProcessing =
+      errorOrUpsertedUserDataProcessing.value;
+    const errorOrMaybeRetrievedProfile = await profileModel.findOneProfileByFiscalCode(
```
Comment (gunzip) [inline UpsertUserDataProcessing/handler.ts:120]: Since this is a web controller it's better to mode `findOneProfileByFiscalCode` into the activity to leave the controller responsive / fast. is it possible ?

File: UpsertedUserDataProcessingOrchestrator/handler.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  // Get and decode orchestrator input
+  const input = context.df.getInput();
+  const errorOrUpsertedUserDataProcessingOrchestratorInput = OrchestratorInput.decode(
+    input
+  );
+
+  if (isLeft(errorOrUpsertedUserDataProcessingOrchestratorInput)) {
+    context.log.error(
+      `${logPrefix}|Error decoding input|ERROR=${readableReport(
+        errorOrUpsertedUserDataProcessingOrchestratorInput.value
```
Comment (gunzip) [inline UpsertedUserDataProcessingOrchestrator/handler.ts:44]: put details only into `context.log.verbose` calls

File: UpsertedUserDataProcessingOrchestrator/handler.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+  const upsertedUserDataProcessingOrchestratorInput =
+    errorOrUpsertedUserDataProcessingOrchestratorInput.value;
+
+  // Log the input
+  context.log.verbose(
+    `${logPrefix}|INPUT=${JSON.stringify(
+      upsertedUserDataProcessingOrchestratorInput
+    )}`
+  );
```
Comment (gunzip) [inline UpsertedUserDataProcessingOrchestrator/handler.ts:58]: do we need this ?

File: SendUserDataProcessingEmailActivity/handler.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    text: emailText,
+    to
+  });
+
+  if (isLeft(errorOrSentMessageInfo)) {
+    const error = Error(
+      `${logPrefix}|Error sending validation email|ERROR=${errorOrSentMessageInfo.value.message}`
+    );
+    context.log.error(error.message);
+    throw error;
```
[the remark this answers is not in the document]
Comment (AleDore) [reply SendUserDataProcessingEmailActivity/handler.ts:87]: It depends on what kind of error we receive, because if it's a temporary network's issue, could be probably useful to retry. In other cases I don't think so...

[the remark this answers is not in the document]
Comment (gunzip) [reply SendUserDataProcessingEmailActivity/handler.ts:87]: can we tell if it's temporary or not ?

[the remark this answers is not in the document]
Comment (AleDore) [reply SendUserDataProcessingEmailActivity/handler.ts:87]: I think that, from a function perspective, we don' t have elements to determine if a network issue is temporary or not.

[the remark this answers is not in the document]
Comment (AleDore) [reply SendUserDataProcessingEmailActivity/handler.ts:87]: I've deleted error throwing. Mark as resolved

File: SendUserDataProcessingEmailActivity/index.ts [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+// Whether we're in a production environment
+const isProduction = process.env.NODE_ENV === "production";
+
+// Mailup
+const mailupUsername = getRequiredStringEnv("MAILUP_USERNAME");
+const mailupSecret = getRequiredStringEnv("MAILUP_SECRET");
+
+// Email data
+const mailFrom = getRequiredStringEnv("DPO_EMAIL_FROM_ADDRESS");
```
Comment (gunzip) [inline SendUserDataProcessingEmailActivity/index.ts:39]: sorry I've changed my mind about this and it's ok to use the global mail_from setting (it's no-reply@io.italia.it)
:))

```suggestion
const mailFrom = getRequiredStringEnv("MAIL_FROM");
```

Comment (AleDore) [reply SendUserDataProcessingEmailActivity/index.ts:39]: Ok let's refactoring!

File: SendUserDataProcessingEmailActivity/handler.ts [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+      // choice to download or delete his own private data stored by the platform
+      const errorOrSentMessageInfo = await sendMail(mailerTransporter, {
+        from,
+        subject,
+        text: emailText,
+        to
+      });
+
+      if (isLeft(errorOrSentMessageInfo)) {
+        const error = Error(
```
Comment (gunzip) [inline SendUserDataProcessingEmailActivity/handler.ts:96]: it looks like this error isn't used anymore so we can keep just the message string

File: SendUserDataProcessingEmailActivity/handler.ts [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        kind: "SUCCESS"
+      });
+    }
+  } else {
+    context.log.error(
+      `${logPrefix}|Error retrieving user's profile|ERROR=${errorOrMaybeRetrievedProfile.value}`
+    );
+    return ActivityResultFailure.encode({
+      kind: "FAILURE",
+      reason: "Error while retrieving user' s profile"
```
Comment (gunzip) [inline SendUserDataProcessingEmailActivity/handler.ts:116]:
```suggestion
      reason: "Error while retrieving user's profile"
```

File: SendUserDataProcessingEmailActivity/handler.ts [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+  );
+  if (isRight(errorOrMaybeRetrievedProfile)) {
+    const maybeRetrievedProfile = errorOrMaybeRetrievedProfile.value;
+    if (isSome(maybeRetrievedProfile)) {
+      const { from, to } = emailDefaults;
+      const subject = `IO - Richiesta di tipo ${choice} da parte di ${fiscalCode}`;
+      const userEmailAddress = maybeRetrievedProfile.value.email;
+      // prepare the text version of the message
+      const emailText = `Un utente di IO ha inoltrato una nuova richiesta:
+  tipo richiesta: ${choice.toString()}
```
Comment (gunzip) [inline SendUserDataProcessingEmailActivity/handler.ts:82]:
```suggestion
  tipo richiesta: ${choice}
```

File: UpsertedUserDataProcessingOrchestrator/handler.ts [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    );
+    return false;
+  }
+
+  const upsertedUserDataProcessingOrchestratorInput =
+    errorOrUpsertedUserDataProcessingOrchestratorInput.value;
+
+  yield context.df.callActivity("SendUserDataProcessingEmailActivity", {
+    ...upsertedUserDataProcessingOrchestratorInput
+  });
```
Comment (gunzip) [inline UpsertedUserDataProcessingOrchestrator/handler.ts:54]:
```suggestion
  yield context.df.callActivity("SendUserDataProcessingEmailActivity", 
  upsertedUserDataProcessingOrchestratorInput);
```

File: UpsertedUserDataProcessingOrchestrator/handler.ts [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+});
+
+export type OrchestratorInput = t.TypeOf<typeof OrchestratorInput>;
+
+export const handler = function*(
+  context: IFunctionContext
+): IterableIterator<unknown> {
+  const logPrefix = `UpsertedUserDataProcessingOrchestrator`;
+
+  const retryOptions = new df.RetryOptions(5000, 10);
```
Comment (gunzip) [inline UpsertedUserDataProcessingOrchestrator/handler.ts:29]: in which case do we have retries ?

**Comment by gunzip** [comment]
can you please fix tests ?

File: UpsertedUserDataProcessingOrchestrator/handler.ts (hunk 16 above)
Comment (AleDore) [reply UpsertedUserDataProcessingOrchestrator/handler.ts:29]: It's a misprint.

File: StoreSpidLogs/__test__/index.test.ts [hunk 17]
```diff
@@ -13,9 +12,8 @@ import { IPString } from "italia-ts-commons/lib/strings";
 import { aFiscalCode } from "../../__mocks__/mocks";
 import { index, SpidBlobItem, SpidMsgItem } from "../index";
 
-const today = format(new Date(), "YYYY-MM-DD");
```
Comment (gunzip) [inline StoreSpidLogs/__test__/index.test.ts:16]: why is this change necessary here ?

Comment (AleDore) [reply StoreSpidLogs/__test__/index.test.ts:16]: It's an optimization, trying to not repeat `new Date()` each time we have to instantiate a Date for testing purposes

Comment (gunzip) [reply StoreSpidLogs/__test__/index.test.ts:16]: I just see a line swapped (I don't see any optimization in this diff) 
anyway it's out of the scope of this pr, can we revert this file to the version from master ?

Comment (AleDore) [reply StoreSpidLogs/__test__/index.test.ts:16]: done!

File: UpsertedUserDataProcessingOrchestrator/handler.ts [hunk 18]
```diff
@@ -46,7 +45,7 @@ export const handler = function*(
     errorOrUpsertedUserDataProcessingOrchestratorInput.value;
 
   yield context.df.callActivity("SendUserDataProcessingEmailActivity", {
-    upsertedUserDataProcessingOrchestratorInput
+    ...upsertedUserDataProcessingOrchestratorInput
```
Comment (AleDore) [inline UpsertedUserDataProcessingOrchestrator/handler.ts:48]: I have to reintroduce this due to ActivityInput compatibility. It expects a Json with this structure: `{"choice": "xxxx",
          "fiscalCode": "xxxxx"}`, while removing `...` we got `{
       "upsertedUserDataProcessingOrchestratorInput": {
          "choice": "XXXXX",
          "fiscalCode": "XXXX",
       },
      }`

File: UpsertedUserDataProcessingOrchestrator/handler.ts [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+      )}`
+    );
+    return false;
+  }
+
+  const upsertedUserDataProcessingOrchestratorInput =
+    errorOrUpsertedUserDataProcessingOrchestratorInput.value;
+
+  yield context.df.callActivity("SendUserDataProcessingEmailActivity", {
+    ...upsertedUserDataProcessingOrchestratorInput
```
[the remark this answers is not in the document]
Comment (AleDore) [reply UpsertedUserDataProcessingOrchestrator/handler.ts:48]: see my comment below

File: UpsertedUserDataProcessingOrchestrator/handler.ts (hunk 18 above)
Comment (gunzip) [reply UpsertedUserDataProcessingOrchestrator/handler.ts:48]: not if you get rid of the brackets as well. 

`f({ ...obj })` equals `f(obj)`  there's no need to spread the object

File: StoreSpidLogs/__test__/index.test.ts [hunk 20]
```diff
@@ -1,6 +1,5 @@
 /* tslint:disable:no-any */
 /* tslint:disable:no-object-mutation */
-
```
Comment (gunzip) [inline StoreSpidLogs/__test__/index.test.ts:3]: can we revert this file to the master one ?
